### PR TITLE
fix: update graphql dependencies to work with Parse Dashboard

### PR DESCRIPTION
**Pull request by FransGH**
Fix for #7650

### New Pull Request Checklist

- [X] I am not disclosing a [vulnerability](https://github.com/parse-community/parse-server/blob/master/SECURITY.md).
- [X] I am creating this PR in reference to an [issue](https://github.com/parse-community/parse-server/issues?q=is%3Aissue).

### Issue Description
Parse Server 4.10.4 installs multiple graphql modules
Related issue:  #7650

### Approach
Updated dependencies in package.json with smallest possible to change to resolve the issue:
"graphql": "15.5.0",          // from "15.4.0"
"graphql-relay": "0.7.0",  // from "0.6.0"

### TODOs before merging
- [x] A changelog entry is created automatically using the pull request title (do not manually add a changelog entry)

### CODE REVIEW

**Comment by mtrezza** [comment]
First, congratulations on your first PR! It seems that you only changed the version numbers in `package.json`. That is not enough as this does not update the `package-lock.json`, see the [build error](https://github.com/parse-community/parse-server/runs/4023004761?check_suite_focus=true):

> npm ERR! cipm can only install packages when your package.json and package-lock.json or npm-shrinkwrap.json are in sync. Please update your lock file with `npm install` before continuing.

To update the packages, revert any changes and run the following:

```sh
npm i -E graphql@15.5.0 graphql-relay@0.7.0
```

Note that dashboard is already on `graphql@15.5.1` I think, so you may want to use that version.

Once you made the changes, you can run the tests locally with `npm test` to see whether all tests pass or other changes are needed. Then we will run the test here with many different Node and MongoDB versions as a final check. Hope this gives some guidance and let me now if you have any questions.

**Comment by FransGH** [comment]
Still need to run the test myself as they fail to run on my dev machine with macOS. Pulling my fork/branch now on a clean docker image..

**Comment by FransGH** [comment]
Did a fresh pull / `npm ci `of "release-4.x.x" within a ubuntu@latest docker, getting the below error on `npm test`.
(same problem on my branch, on the 'release' branch I get a different error:  can't find module <...>/SchemaCache)

Do I need to install other dependencies to run the tests?

```
root@d14a71f45616:/parse-server# npm test

> parse-server@4.10.4 pretest /parse-server
> cross-env MONGODB_VERSION=${MONGODB_VERSION:=4.0.4} MONGODB_TOPOLOGY=${MONGODB_TOPOLOGY:=standalone} MONGODB_STORAGE_ENGINE=${MONGODB_STORAGE_ENGINE:=mmapv1} mongodb-runner start

  ◜ Starting a MongoDB deployment to test against...
> parse-server@4.10.4 test /parse-server
> npm run testonly


> parse-server@4.10.4 testonly /parse-server
> cross-env MONGODB_VERSION=${MONGODB_VERSION:=4.0.4} MONGODB_TOPOLOGY=${MONGODB_TOPOLOGY:=standalone} MONGODB_STORAGE_ENGINE=${MONGODB_STORAGE_ENGINE:=mmapv1} TESTING=1 jasmine

/parse-server/spec/ParseLiveQuery.spec.js:952
      expect(prev?.sessionToken).toBeUndefined();
                  ^

SyntaxError: Unexpected token .
    at Module._compile (internal/modules/cjs/loader.js:723:23)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:789:10)
    at Module.load (internal/modules/cjs/loader.js:653:32)
    at tryModuleLoad (internal/modules/cjs/loader.js:593:12)
    at Function.Module._load (internal/modules/cjs/loader.js:585:3)
    at Module.require (internal/modules/cjs/loader.js:692:17)
    at require (internal/modules/cjs/helpers.js:25:18)
    at /parse-server/node_modules/jasmine/lib/jasmine.js:89:5
    at Array.forEach (<anonymous>)
    at Jasmine.loadSpecs (/parse-server/node_modules/jasmine/lib/jasmine.js:88:18)
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! parse-server@4.10.4 testonly: `cross-env MONGODB_VERSION=${MONGODB_VERSION:=4.0.4} MONGODB_TOPOLOGY=${MONGODB_TOPOLOGY:=standalone} MONGODB_STORAGE_ENGINE=${MONGODB_STORAGE_ENGINE:=mmapv1} TESTING=1 jasmine`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the parse-server@4.10.4 testonly script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /root/.npm/_logs/2021-10-27T14_40_21_215Z-debug.log
npm ERR! Test failed.  See above for more details.
root@d14a71f45616:/parse-server# 
```

**Comment by mtrezza** [comment]
Which node version are you using?

**Comment by FransGH** [comment]
I've tried again with the same node/npm versions as in the workflow:

```
root@d14a71f45616:/parse-server# npm -v
6.14.12
root@d14a71f45616:/parse-server# node -v
v10.24.1
```

When I do `npm run coverage ` on either on  "release-4.x.x" or "tags/4.10.4" it always fails on 

```
/parse-server/spec/ParseLiveQuery.spec.js:952
      expect(prev?.sessionToken).toBeUndefined();

```

I'll try to troubleshoot this a bit further. Guess it must be some dependency or could it be that the commits I'm using to test have an issue?

**Comment by mtrezza** [comment]
Can you try to use a higher node version? I suspect optional chaining (`optional?.key`) does not exist in Node 10. The tests are not babel compiled, so if a test uses an unsupported syntax, it won't run.

**Comment by mtrezza** [comment]
You can run the test in Node 10, but change test to not use optional chaining.

File: package.json
Comment (mtrezza) [inline package.json]: please revert this, we will try to automatically bump this via release automation

**Comment by mtrezza** [comment]
The test fail, luckily the test logs output on which line. I suggest running this locally for debugging to see what exactly causes the issue at that line:

> Error: Object not found.
>    at then.then.result (/home/runner/work/parse-server/parse-server/lib/Controllers/DatabaseController.js:49:3084)
>    at process._tickCallback (internal/process/next_tick.js:68:7)

It's the same error as in https://github.com/parse-community/parse-server/pull/7599. That means the CI on that branch is currently broken. I opened https://github.com/parse-community/parse-server/issues/7659 for this. If you find a solution to this, we can open a separate PR for a fix. You could just make a new branch on 4.x without any changes and try to run the tests locally on Node 10 to investigate. We can then follow up the discussion there. After that fix, this PR can be merged.

**Comment by FransGH** [comment]
I left that one commit from 7599 out as I was unclear what it did and like to keep the PR to an absolute minimum. The other two commits were just syntax issues. 

On the "object not found" output: this is not an error but seem to be 'expected' trace output from [EnableExpressErrorHandler.spec.js](https://github.com/parse-community/parse-server/blob/alpha/spec/EnableExpressErrorHandler.spec.js). I see this too when running the test locally and it's not an issue.

Don't know at the moment why the tests again fail, I'll have a closer look...

**Comment by FransGH** [comment]
There seem to be 15min timeouts within the latest [CI workflows](https://github.com/parse-community/parse-server/blob/alpha/.github/workflows/ci.yml). 

My guess is that GitHub Actions always triggers the latest revision of the workflow file, causing ci.yml from the alpha branch to precede over the release-4.x.x branch.

Could you maybe increase the 'matrix' timeouts to 30min in the alpha branch and run the workflows again?
They should than either run trough or get killed after 30min...

**Comment by mtrezza** [comment]
The workflow that runs in the one that is in the brach that triggered the workflow, so the workflow file is correct. They did not time out, I stopped the CI because it stalled. A test should not take longer than 15 mins, more like <10 mins.

If you run npm test locally on npm 10, do they all pass?

**Comment by FransGH** [comment]
I also canceled after a longer time before doing the last commit. It didn't look stalled and was just taking a very long to complete. Will now let it run overnight...

What I don't understand is how 4.10.4 got released when the ci workflow seems to be broke? 

It looks like you are using a different CI workflow in your [fork](https://github.com/mtrezza/parse-server/actions/workflows/ci.yml)? Since this seems to work, wouldn't it make sense to merge that into into a 4.10.x PR?

image

**Comment by FransGH** [comment]
Ok, found the failed test, it's related to this https://github.com/parse-community/parse-server/security/advisories/GHSA-7pr3-p5fm-8r9x.

The new test "should strip out session token in LiveQuery" fails and is included in both ParseLiveQuery.spec.js and ParseUser.spec.js. When I comment it out, the test complete ok on node 10. I'll try to fix this test and add it to the PR...

**Comment by mtrezza** [comment]
Amazing! Can you please open a separate PR for the fix and reference issue https://github.com/parse-community/parse-server/issues/7659?

**Comment by FransGH** [comment]
@mtrezza, sorry, wanted to create a new branch/pr for this but mixed something up. Maybe just let the workflows run to see if the issue is really fixed? Happy to create a separate PR for this if the CI issue is solved..

**Comment by mtrezza** [comment]
Workflows running

**Comment by mtrezza** [comment]
The CI has passed, well done! Could you extract the CI fix into a separate PR so we can evaluate that separately?

**Comment by FransGH** [comment]
New PR #7661 for CI fix created. What do we do with this one?

**Comment by mtrezza** [comment]
This one is for the graphql upgrade as I understand? So let's first merge the CI fix, then we can merge this one.

**Comment by mtrezza** [comment]
@FransGH Could you merge the branch into this PR, so we can see whether the CI passes? https://github.com/parse-community/parse-server/pull/7661 has been merged.

**Comment by mtrezza** [comment]
There seems to be a mix up with the other PR that optimizes the CI time?  Could you reduce this PR to only upgrade the GraphQL dependencies?

**Comment by FransGH** [comment]
Sorry for that, fix it!

File: package.json
Comment (mtrezza) [inline package.json]: Strange that this shows up here as a change, didn't we just add this with the previous PR? I think this branch is not fully up-to-date.

Comment (FransGH) [reply package.json]: I merged fix-lts-ci-7659 into fix-graphql-deps. Did a reset on fix-lts-ci-7659 to remove my latest changes. Isn't `"jasmine-spec-reporter": "7.0.0"` part of fix-lts-ci-7659? (ps: are you you from Vienna? I'm in Mauer, 1230...)

Comment (FransGH) [reply package.json]: Oh, now I'm confused, where did the graphql changes go???

Comment (mtrezza) [reply package.json]: Maybe you want to force push this branch to the latest commit on the 4.x branch and re-do the graphql changes, if they were not too many?

Comment (FransGH) [reply package.json]: ok, yeah that might be best!

**Comment by FransGH** [comment]
Ok, 1st time working from a fork, doing a PR and discovering 'fetch upstream'.. Now things should be ok again :)

**Comment by mtrezza** [comment]
You may find "GitHub Desktop" helpful that gives you a visual UI for the usual operations needed for contributing.

The next dashboard release will use `"graphql": "15.6.1"` while this PR upgrades to `"graphql": "15.5.1"`. Will that cause the same error, or was the error only because of a major version difference? If the graphql versions have to be completely identical, then we may have to open an issue with the maintainers of the package that is throwing that error.

**Comment by mtrezza** [comment]
Note: before merging this; auto-release needs to be configured for the `release-4.x.x` branch.
https://github.com/parse-community/parse-server/issues/7674

**Comment by mtrezza** [comment]
Auto-release has been added to `release-4.x.x` branch, so this can be merged if it passes review.

**Review by mtrezza — APPROVED** [review]

**Comment by parseplatformorg** [comment]
🎉 This change has been released in version [4.10.6](https://github.com/parse-community/parse-server/releases/tag/4.10.6)

**Pull request merged**
